File: apple/Sources/CCDAEngine/Models/Entry/CCDAEntryType.swift
import Foundation

/// XML element type for a structured C-CDA entry.
public enum CCDAEntryType: Hashable, Sendable {
    /// CDA act entry.
    case act
    /// CDA encounter entry.
    case encounter
    /// CDA observation entry.
    case observation
    /// CDA organizer entry.
    case organizer
    /// CDA procedure entry.
    case procedure
    /// CDA substanceAdministration entry.
    case substanceAdministration
    /// CDA supply entry.
    case supply
    /// Unsupported entry element name preserved from source XML.
    case unsupported(String)

    /// Creates an entry type from an XML element name.
    public init(elementName: String) {
        switch elementName {
        case "act":
            self = .act
        case "encounter":
            self = .encounter
        case "observation":
            self = .observation
        case "organizer":
            self = .organizer
        case "procedure":
            self = .procedure
        case "substanceAdministration":
            self = .substanceAdministration
        case "supply":
            self = .supply
        default:
            self = .unsupported(elementName)
        }
    }

    /// Original XML element name.
    public var rawValue: String {
        switch self {
        case .act:
            return "act"
        case .encounter:
            return "encounter"
        case .observation:
            return "observation"
        case .organizer:
            return "organizer"
        case .procedure:
            return "procedure"
        case .substanceAdministration:
            return "substanceAdministration"
        case .supply:
            return "supply"
        case .unsupported(let rawValue):
            return rawValue
        }
    }
}

File: android/ccda-engine/src/main/kotlin/com/ccdakit/engine/Models/Entry/CCDAEntryType.kt
package com.ccdakit.engine

/** XML element type for a structured C-CDA entry. */
sealed class CCDAEntryType(open val rawValue: String) {
    /** CDA act entry. */
    data object Act : CCDAEntryType("act")
    /** CDA encounter entry. */
    data object Encounter : CCDAEntryType("encounter")
    /** CDA observation entry. */
    data object Observation : CCDAEntryType("observation")
    /** CDA organizer entry. */
    data object Organizer : CCDAEntryType("organizer")
    /** CDA procedure entry. */
    data object Procedure : CCDAEntryType("procedure")
    /** CDA substanceAdministration entry. */
    data object SubstanceAdministration : CCDAEntryType("substanceAdministration")
    /** CDA supply entry. */
    data object Supply : CCDAEntryType("supply")
    /** Unsupported entry element name preserved from source XML. */
    data class Unsupported(override val rawValue: String) : CCDAEntryType(rawValue)

    companion object {
        /** Creates an entry type from an XML element name. */
        fun fromElementName(name: String): CCDAEntryType = when (name) {
            "act" -> Act
            "encounter" -> Encounter
            "observation" -> Observation
            "organizer" -> Organizer
            "procedure" -> Procedure
            "substanceAdministration" -> SubstanceAdministration
            "supply" -> Supply
            else -> Unsupported(name)
        }
    }
}
